Add support for marking doctest items as distinct from normal code

Nashenas88 opened the pull request:
This adds `HighlightTag::Generic | HighlightModifier::Injected` as the default highlight for all elements within a doctest. Please feel free to suggest that a new tag be created or a different one used.

![Screenshot from 2020-06-23 09-18-13](https://user-images.githubusercontent.com/1673130/85408493-9752ce00-b532-11ea-94fe-197353ccc778.png)

Fixes #4929 
Fixes #4939

ltentrup commented:
As a bonus, one could change the opacity of the injections CSS class to implement the dimming in the HTML output

Nashenas88 commented:
@ltentrup I did italic and underline since I was testing within vs code. Dimming is my preference but to do that in vs code, I'd have to brute force override every single highlighting rule (since opacity is only exposed through the RGBA of the foreground color), and this was quicker to do :sweat_smile: .

ltentrup commented:
Sure, I meant the HTML output in syntax_highlighting/html.rs 😅

Nashenas88 commented:
Oh you mean the default html highlighting? I can definitely add that in.

Nashenas88 commented:
This is what gets generated now in the doc tests. I also added one for documentation, feel free to suggest different css for either.

![Screenshot from 2020-06-18 10-31-08](https://user-images.githubusercontent.com/1673130/85033384-e07bda00-b14e-11ea-883e-5038b35bb832.png)

ltentrup commented:
I like the dimmed style but I think there is still a problem where previously highlighted tokens are no longer highlighted

ltentrup commented:
I think I know where the problem comes from: if we have a parse tree `OUTER[INNER[]]` with the same range and OUTER was highlighted before (but not INNER), with the change INNER overwrites OUTER (with operator injection semantic token).
I propose a different solution: instead of modifying the syntax highlighting, I would modify the injection logic: instead of deleting the parent range during intersection, change the intersected parent range to the operator injection semantic token.

matklad commented:
bors d=@ltentrup 

:)

ltentrup commented:
Thanks @matklad :) 

@Nashenas88  here is the relevant code:
https://github.com/rust-analyzer/rust-analyzer/blob/ec6df5d3e866c215f7e699366ba55a7112ab4149/crates/ra_ide/src/syntax_highlighting.rs#L344-L357

After the intersection in line 347, the type of the splitted parent tokens `prev[idx]` and `cloned` have to be changed (and `delete` has to be changed to the overwriting semantic type)

Nashenas88 commented:
@ltentrup it's probably because it's been a long day, but I couldn't get that change working. I might have time Sunday night (I'm in the US Eastern time zone), but if that doesn't work, would you be available to chat on zulip sometime Monday?

ltentrup commented:
@Nashenas88 you can try to ping me but I can only work on ra in my free time. I can have a look at a WIP branch, though, or try to answer specific questions.

Nashenas88 commented:
I just pushed something that "fixes" the highlighting for most cases, by adding a new punctuation semantic token that is only emitted for doctests. I think though this is still subpar compared to your idea, and now that I've been messed around with this enough I understand what you mean. I'll attempt to tackle this again after work.

ltentrup commented:
I've just seen your message, let me know if there are remaining issues

Nashenas88 commented:
@ltentrup sorry for the delay. I ran into #5002 while debugging this and decided to fix that first.

I now have a good understanding of how this merging logic works. What's not clear is which tag to use for parts of the ranges that are not matched to other children in the loop. Below is some of the code I was debugging, and below that the debug output I was working with:
```rust
/// ```
/// assert!(foo.bar());
/// ```
pub const fn abc() {}
```
<details>
<summary>Click to expand debug output</summary>

```
[crates/ra_ide/src/syntax_highlighting/injection.rs:150] text = "fn doctest() {\nassert!(foo.bar());}\n"

Child: HighlightedRange { range: 8..12, highlight: Highlight { tag: Comment, modifiers: HighlightModifiers(8) }, binding_hash: None }
Matched to HighlightedRange { range: 8..31, highlight: Highlight { tag: Comment, modifiers: HighlightModifiers(8) }, binding_hash: None }
Split to:
Cloned HighlightedRange { range: 12..31, highlight: Highlight { tag: Comment, modifiers: HighlightModifiers(8) }, binding_hash: None }
prev   HighlightedRange { range: 8..8, highlight: Highlight { tag: Comment, modifiers: HighlightModifiers(8) }, binding_hash: None }

Child: HighlightedRange { range: 12..19, highlight: Highlight { tag: Macro, modifiers: HighlightModifiers(16) }, binding_hash: None }
Not matched, inserting

Child: HighlightedRange { range: 19..20, highlight: Highlight { tag: Punctuation, modifiers: HighlightModifiers(16) }, binding_hash: None }
Not matched, inserting

Child: HighlightedRange { range: 23..24, highlight: Highlight { tag: Punctuation, modifiers: HighlightModifiers(16) }, binding_hash: None }
Not matched, inserting

Child: HighlightedRange { range: 27..28, highlight: Highlight { tag: Punctuation, modifiers: HighlightModifiers(16) }, binding_hash: None }
Not matched, inserting

Child: HighlightedRange { range: 28..29, highlight: Highlight { tag: Punctuation, modifiers: HighlightModifiers(16) }, binding_hash: None }
Not matched, inserting

Child: HighlightedRange { range: 29..30, highlight: Highlight { tag: Punctuation, modifiers: HighlightModifiers(16) }, binding_hash: None }
Not matched, inserting

Child: HighlightedRange { range: 30..31, highlight: Highlight { tag: Punctuation, modifiers: HighlightModifiers(16) }, binding_hash: None }
Not matched, inserting

Child: HighlightedRange { range: 31..32, highlight: Highlight { tag: Punctuation, modifiers: HighlightModifiers(16) }, binding_hash: None }
```

</details>

So it looks like we should take advantage of `cloned` because it contains the exact range that we want the `Injected` modifier on. The issue is that `Injected` is only a modifier, and not a tag. So, as we go through the rest of the children on that line, we can fill in most of the line, but there are still items that get missed. For example, the ranges 20..23 and 24..27 do not get highlighted (In this case they resolve to `foo` and `bar`). It's not clear whether to tag those with `HighlightTag::Operator` or include a new tag for `highlight_element` to emit. The other thought that comes up is this, if we can generate a tag for every item in the comment block, why not put out effort in `highlight_element` instead to make sure everything gets a `HighlightTag`? I've already modified it so these excessive ranges only get emitted within doctests and not through the rest of the code.

Nashenas88 commented:
@ltentrup I think we got everything now! Thanks for helping to push this to the clearly better solution. Below is what it currently looks like. I also removed the `Punctuation` tag and replaced it with a `Generic` tag. Please feel free to suggest a better name (I'm usually terrible at naming thing), its intent is only for items in doctests that are normally not highlighted in regular code.

![Screenshot from 2020-06-23 09-18-13](https://user-images.githubusercontent.com/1673130/85408493-9752ce00-b532-11ea-94fe-197353ccc778.png)

Nashenas88 commented on crates/ra_ide/src/syntax_highlighting.rs:
```diff
@@ -328,12 +331,12 @@ impl HighlightedRangeStack {
+    fn pop_and_inject(&mut self, inject: bool) {
         let mut children = self.stack.pop().unwrap();
         let prev = self.stack.last_mut().unwrap();
         children.sort_by_key(|range| range.range.start());
```
Had to use `Option<Highlight>` because we needed the `Injected` modifier.

Nashenas88 commented on crates/ra_ide/src/snapshots/highlight_doctest.html:
````diff
@@ -32,65 +34,65 @@
     <span class="comment documentation">/// # Examples</span>
     <span class="comment documentation">///</span>
     <span class="comment documentation">/// ```</span>
-    <span class="comment documentation">/// #</span> <span class="attribute">#![</span><span class="function attribute">allow</span><span class="attribute">(unused_mut)]</span>
````
Fixed with the latest change.

Nashenas88 commented on crates/ra_ide/src/snapshots/highlight_doctest.html:
````diff
@@ -44,9 +44,9 @@
-    <span class="comment documentation">/// #</span> <span class="attribute">#![</span><span class="function attribute">allow</span><span class="attribute">(unused_mut)]</span>
-    <span class="comment documentation">/// </span><span class="keyword">let</span> <span class="keyword">mut</span> <span class="variable declaration mutable">foo</span>: <span class="struct">Foo</span> = <span class="struct">Foo</span>::<span class="function">new</span>();
-    <span class="comment documentation">/// ```</span>
+    <span class="comment documentation">/// #</span> <span class="operator injected">#</span><span class="operator injected">!</span><span class="operator injected">[</span><span class="operator injected">allow</span><span class="operator injected">(</span><span class="operator injected">unused_mut</span><span class="operator injected">)</span><span class="operator injected">]</span><span class=" ...
````
Fixed with the latest change.

ltentrup commented on crates/ra_ide/src/syntax_highlighting.rs:
```diff
@@ -514,13 +540,13 @@ fn highlight_element(
             let expr = prefix_expr.expr()?;
             let ty = sema.type_of_expr(&expr)?;
+            let mut h = HighlightTag::Operator.into();
             if !ty.is_raw_ptr() {
-                return None;
+                h
+            } else {
+                h |= HighlightModifier::Unsafe;
+                h
```
Is this change still necessary?

Nashenas88 replied:
I wouldn't say it's unnecessary, but it is unrelated to this change. Should I just remove it? It could always be added as a separate PR.

Nashenas88 replied:
For context, this was just ensuring that safe deref operations are marked as an operation. The only other items with this currently are the question mark operator and deref when the operation is unsafe.

ltentrup commented on crates/ra_ide/src/syntax_highlighting.rs:
```diff
@@ -324,16 +324,27 @@ impl HighlightedRangeStack {
     /// Can be used to do injections that span multiple ranges, like the
     /// doctest injection below.
-    /// If `delete` is set to true, the parent range is deleted instead of
+    /// If `inject` is set to true, the parent range is deleted instead of
     /// intersected.
```
```suggestion
    /// If `overwrite_parent` is non-optional, the highlighting of the parent range
    /// is overwritten with the argument.
```

ltentrup commented on crates/ra_ide/src/syntax_highlighting.rs:
```diff
@@ -343,26 +354,41 @@ impl HighlightedRangeStack {
-                    .binary_search_by_key(&child.range.start(), |range| range.range.start())
-                    .unwrap_or_else(|x| x);
-                prev.insert(idx, child);
+                let maybe_idx =
+                    prev.iter().position(|parent| parent.range.contains(child.range.start()));
+                if let (Some(_), Some(idx)) = (overwrite_parent, maybe_idx) {
```
Have you observed this situation? I am pretty sure this can't happen in the current form. If a child-range only partially overlaps a parent range this has to be a bug

Nashenas88 replied:
I ran into this after ensuring `cloned` was added into `prev`. This was specifically the case with the mult-iline comment. I'm not sure if it's that multi-line comments also include the newline while single-line comments don't. Before this change, the child would be completely outside of the parent range, so it was possible to just add it to `prev`. Now because we're tracking the `cloned` area from the original doc-comment split, there's a mismatch at the very end (`cloned` from the original split is shorter than `child`).

ltentrup replied:
Alright, can you transform the if into a match? That would probably make it a bit more readable

Nashenas88 replied:
Done

ltentrup commented:
bors r+
